File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/imagegenerator/SnapshotActivity.java
package com.mapbox.mapboxsdk.testapp.activity.imagegenerator;

import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;

import com.mapbox.mapboxsdk.constants.Style;
import com.mapbox.mapboxsdk.maps.MapView;
import com.mapbox.mapboxsdk.maps.MapboxMap;
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback;
import com.mapbox.mapboxsdk.testapp.R;

import java.util.Locale;

/**
 * Test activity showcasing the Snapshot API to create and display a bitmap of the current shown Map.
 */
public class SnapshotActivity extends AppCompatActivity implements OnMapReadyCallback, View.OnClickListener {

  private MapView mapView;
  private MapboxMap mapboxMap;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_snapshot);

    mapView = (MapView) findViewById(R.id.mapView);
    mapView.onCreate(savedInstanceState);
    mapView.getMapAsync(this);
  }

  @Override
  public void onMapReady(MapboxMap map) {
    mapboxMap = map;
    mapboxMap.setStyleUrl(Style.OUTDOORS);
    FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
    if (fab != null) {
      fab.setColorFilter(ContextCompat.getColor(SnapshotActivity.this, R.color.primary));
      fab.setOnClickListener(this);
    }
  }

  @Override
  public void onClick(View view) {
    final long startTime = System.nanoTime();
    mapboxMap.snapshot(snapshot -> {
      long endTime = System.nanoTime();
      long duration = (long) ((endTime - startTime) / 1e6);
      ImageView snapshotView = (ImageView) findViewById(R.id.imageView);
      snapshotView.setImageBitmap(snapshot);
      Toast.makeText(
        SnapshotActivity.this,
        String.format(Locale.getDefault(), "Snapshot taken in %d ms", duration),
        Toast.LENGTH_LONG).show();
    });
  }

  @Override
  protected void onStart() {
    super.onStart();
    mapView.onStart();
  }

  @Override
  protected void onResume() {
    super.onResume();
    mapView.onResume();
  }

  @Override
  protected void onPause() {
    super.onPause();
    mapView.onPause();
  }

  @Override
  protected void onStop() {
    super.onStop();
    mapView.onStop();
  }

  @Override
  public void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);
    mapView.onSaveInstanceState(outState);
  }

  @Override
  public void onLowMemory() {
    super.onLowMemory();
    mapView.onLowMemory();
  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    mapView.onDestroy();
  }
}

File: platform/android/MapboxGLAndroidSDKTestApp/src/main/java/com/mapbox/mapboxsdk/testapp/activity/imagegenerator/SnapshotActivity.kt
package com.mapbox.mapboxsdk.testapp.activity.imagegenerator

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import com.mapbox.mapboxsdk.log.Logger
import com.mapbox.mapboxsdk.maps.MapView
import com.mapbox.mapboxsdk.maps.MapboxMap
import com.mapbox.mapboxsdk.maps.OnMapReadyCallback
import com.mapbox.mapboxsdk.maps.Style
import com.mapbox.mapboxsdk.testapp.R
import kotlinx.android.synthetic.main.activity_snapshot.*
import timber.log.Timber

/**
 * Test activity showcasing the Snapshot API to create and display a bitmap of the current shown Map.
 */
class SnapshotActivity : AppCompatActivity(), OnMapReadyCallback {

  private lateinit var mapboxMap: MapboxMap

  private val idleListener = object : MapView.OnDidFinishRenderingFrameListener {
    override fun onDidFinishRenderingFrame(fully: Boolean) {
      if (fully) {
        mapView.removeOnDidFinishRenderingFrameListener(this)
        Logger.v(TAG, LOG_MESSAGE)
        mapboxMap.snapshot { snapshot ->
          imageView.setImageBitmap(snapshot)
          mapView.addOnDidFinishRenderingFrameListener(this)
        }
      }
    }
  }

  public override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_snapshot)
    mapView.onCreate(savedInstanceState)
    mapView.getMapAsync(this)
  }

  override fun onMapReady(map: MapboxMap) {
    mapboxMap = map
    mapboxMap.setStyle(Style.Builder().fromUri(Style.OUTDOORS)) { mapView.addOnDidFinishRenderingFrameListener(idleListener) }
  }

  override fun onStart() {
    super.onStart()
    mapView.onStart()
  }

  override fun onResume() {
    super.onResume()
    mapView.onResume()
  }

  override fun onPause() {
    super.onPause()
    mapboxMap.snapshot {
      Timber.e("Regression test for https://github.com/mapbox/mapbox-gl-native/pull/11358")
    }
    mapView.onPause()
  }

  override fun onStop() {
    super.onStop()
    mapView.onStop()
  }

  public override fun onSaveInstanceState(outState: Bundle) {
    super.onSaveInstanceState(outState)
    mapView.onSaveInstanceState(outState)
  }

  override fun onLowMemory() {
    super.onLowMemory()
    mapView.onLowMemory()
  }

  public override fun onDestroy() {
    super.onDestroy()
    mapView.removeOnDidFinishRenderingFrameListener(idleListener)
    mapView.onDestroy()
  }

  companion object {
    const val TAG = "Mbgl-SnapshotActivity"
    const val LOG_MESSAGE = "OnSnapshot"
  }
}